lib/index.js

```diff
@@ -1,67 +1,133 @@
+const Stream = require('stream');
 const Hoek = require('@hapi/hoek');
-const debug = require('debug')('hapi:plugins:trace-headers');
+const NodeCache = require('node-cache');
+const debug = require('debug')('hapi:plugins:good-tracer');
 const { v4: uuidv4 } = require('uuid');
 const { get, set, isNaN } = require('lodash');
 
-const defaults = {
+const NAME = 'goodTracer';
+
+const DEFAULTS = {
     traceUUIDHeader: 'x-gg-trace-uuid',
     traceSeqIDHeader: 'x-gg-trace-seqid',
+    cache: {
+        ttl: 2 * 60, // 2 minutes
+        checkPeriod: 5 * 60, // 5 minutes
+        maxKeys: 5000,
+        extendTTLOnGet: true, // extend TTL on cache get
+        useClones: false
+    }
 };
 
-exports.incrementSeqID = ({ request, key }) => {
+const internals = {};
+
+internals.incrementSeqID = ({ request, key }) => {
+    if (get(request, ['headers', key]) === undefined) {
+        set(request, ['plugins', NAME, key], 0);
+        return true;
+    }
+
     const ogSeqId = Number(get(request, ['headers', key], 0));
 
     if (isNaN(ogSeqId)) {
         debug('SeqId is NaN: %o', get(request, ['headers', key]));
+        set(request, ['plugins', NAME, key], get(request, ['headers', key]));
         return true;
     }
 
     debug('SeqId before: %d', ogSeqId);
-    set(request, ['headers', key], ogSeqId + 1);
-    debug('SeqId after: %d', get(request, ['headers', key]));
+    set(request, ['plugins', NAME, key], ogSeqId + 1);
+    debug('SeqId after: %d', get(request, ['plugins', NAME, key]));
     return true;
 };
 
+internals.injectTracerObject = (data) => {
+    debug('event type: %s', data.event);
+    // lookup info from cache
+    const requestID = get(data, 'id');
+    if (requestID) {
+        debug('found request id: %s', requestID);
+        const value = internals.cache.get(requestID);
+        if (value) {
+            debug('found value in cache: %o', value);
+            set(data, 'tracer', value);
+            if (internals.settings.cache.extendTTLOnGet) {
+                // reset TTL on cached value
+                internals.cache.ttl(requestID);
+            }
+        }
+    }
+
+    return data;
+};
+
+internals.GoodTracerStream = class GoodSourceTracer extends Stream.Transform {
+    constructor(options) {
+        super({ objectMode: true });
+
+        this._settings = Hoek.applyToDefaults(internals.settings, options);
+    }
+
+    _transform(data, enc, next) { // eslint-disable-line class-methods-use-this
+        return next(null, internals.injectTracerObject(data));
+    }
+};
+
 exports.register = (server, options) => {
-    const settings = Hoek.applyToDefaults(defaults, options);
-    debug('Initialized with settings: %o', settings);
+    internals.settings = Hoek.applyToDefaults(DEFAULTS, options);
+    debug('Initialized with settings: %o', internals.settings);
+
+    internals.cache = new NodeCache(internals.settings.cache);
```

Comment:
What benefits do we find with having a shared cache between all instances of this plugin? Also, as I read this, we are creating a new cache and storing it at the module level on each registration. This will break the isolation of multi-server instance applications in that they will now all be sharing the same cache instance.

Reply:
womp womp

Reply:
As for the multi server instance applications, that is okay because the `request` is unique to a single instance. The `headers` would be passed to the next instance `request` chain. I don't think we need to be concerned with the multi-server situation as a request cannot be processed across many servers.

Reply:
There is only one registration of the plugin at server startup, unless I am mistaken. We can move the cache creation to a highlevel is that is desired, but I don't think the server would be recreated often.

Reply:
That is fine. A potential race condition could be created with the instantiation of the cache in the server register method. Consider this:

1. Server 1 created
2. Server 1 receives request
3. Server 2 is created during Server 1's handling of on going request
4. Server 1 begins logging request information but is unable to find cached request headers because the `internals.cache` has been reassigned during Server 2 registration

I am fully aware how much of an edge case this is, but it's worth noting it. A fix for this edge case is to make `internals.cache` a singleton at which point the first server's configurable cache settings will be the settings for the entire node application.

Reply:
Thanks for the feedback! edge case or not, moving to a singleton for the cache makes sense. Please see the updates pushed.

Reply:
@mrlannigan Alright, great conversations around this offline. I like where we landed. I have pushed the updates for the factory method to generate the `GoodTracerStream` and dropped the singleton.